src/stream/src/executor/backfill/snapshot_backfill/state.rs

```diff
@@ -0,0 +1,399 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::mem::replace;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use futures::future::try_join_all;
+use futures::TryFutureExt;
+use risingwave_common::bitmap::Bitmap;
+use risingwave_common::hash::{VirtualNode, VnodeBitmapExt};
+use risingwave_common::must_match;
+use risingwave_common::row::{OwnedRow, Row, RowExt};
+use risingwave_common::types::{DataType, ScalarImpl};
+use risingwave_common::util::row_serde::OrderedRowSerde;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(super) enum EpochBackfillProgress {
+    Consuming { latest_pk: OwnedRow },
+    Consumed,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub(super) struct VnodeBackfillProgress {
+    pub(super) epoch: u64,
+    pub(super) row_count: usize,
+    pub(super) progress: EpochBackfillProgress,
+}
+
+/// `vnode`, `epoch`, `row_count`, `is_finished`
+const EXTRA_COLUMN_TYPES: [DataType; 4] = [
+    DataType::Int16,
+    DataType::Int64,
+    DataType::Int64,
+    DataType::Boolean,
+];
+
+impl VnodeBackfillProgress {
+    fn validate_progress_table_schema(
+        progress_table_column_types: &[DataType],
+        upstream_pk_column_types: &[DataType],
+    ) -> StreamExecutorResult<()> {
+        if progress_table_column_types.len()
+            != EXTRA_COLUMN_TYPES.len() + upstream_pk_column_types.len()
+        {
+            return Err(anyhow!(
+                "progress table columns len not matched with the len derived from upstream table pk. progress table: {:?}, pk: {:?}",
+                progress_table_column_types,
+                upstream_pk_column_types)
+                .into()
+            );
+        }
+        for (expected_type, progress_table_type) in EXTRA_COLUMN_TYPES
+            .iter()
+            .chain(upstream_pk_column_types.iter())
+            .zip_eq_debug(progress_table_column_types.iter())
+        {
+            if expected_type != progress_table_type {
+                return Err(anyhow!(
+                    "progress table column not matched with upstream table schema: progress table: {:?}, pk: {:?}",
+                    progress_table_column_types,
+                    upstream_pk_column_types)
+                    .into()
+                );
+            }
+        }
+        Ok(())
+    }
+
+    pub(super) fn from_row(row: &OwnedRow, pk_serde: &OrderedRowSerde) -> Self {
+        assert_eq!(
+            row.len(),
+            pk_serde.get_data_types().len() + EXTRA_COLUMN_TYPES.len() - 1, // Pk not included
```

Comment:
```suggestion
            pk_serde.get_data_types().len() + EXTRA_COLUMN_TYPES.len() - 1, // Vnode not included
```

I suppose you meant vnode. `PK` is rather vague, since it could refer to the vnode-pk for the backfill state table, or the upstream-pk of the upstream table.